### ISSUE TITLE
Understanding authenticate target_url and unauth_action arguments in authenticate function. Is there any documentation for these?

### ISSUE THREAD

**Issue by dcs3spp** [zmartzone/lua-resty-openidc#272]
Hi, 

I am learning how to use lua-resty-openidc and have managed to get a session cookie returned containing access token and id token. So far I am finding the library really useful and considering the possibility of using it as an api proxy/gateway to perform authentication via keycloak before sending the request upstream.

However, I am struggling to understand what the following arguments are used for in the authenticate function:
- target_url: How does this relate to the *redirect_uri* specified in options? Is it the same thing? In the code block below I have set this to *nil*.
- unauth_action:  Looking at the source code, if this is set to "pass" and currently unauthenticated then redirect to the open connect provider. Are there any additional values for unauth_action? 

`require("resty.openidc").authenticate(opts, nil, "unauth_value_other_than_pass", session_opts)`

```
  server {
    listen 80;
    charset utf-8;
    default_type text/html;

    resolver 127.0.0.11 ipv6=off;

    set $session_secret <...secret...>;

    location /auth {
      access_by_lua_block {
        local opts = {
          redirect_uri = "http://angular.example.com:4200",
          discovery = "http://keycloak.example.com:8080/auth/realms/client/.well-known/openid-configuration",
          client_id = "client",
          client_secret = "<secret>",
          ssl_verify = "no",
          accept_none_alg = false,
          accept_unsupported_alg = false,
          renew_access_token_on_expiry = true,
          access_token_expires_in = 3600,
          revoke_tokens_on_logout = true,
          session_contents = {id_token=true, access_token=true},
        }

        local session_opts = { cookie = { domain = ".example.com" } }
        local res, err = require("resty.openidc").authenticate(opts, nil, "donotpass", session_opts)
      }
    }
  }
```

**Comment by zandbelt** [comment]
you hit a fair point about documentation: right now it consists of reading the Lua script code; if anyone can help on this, it would be great...

**Comment by dcs3spp** [comment]
Thanks @zandbelt. Have looked at the source code again for authenticate:
- Does the *target_url* argument relate to the url visited before authentication occurred? not sure upon inspecting the source code that I understand the usage with respect to to this argument....any help/resources/advice with respect to this argument would be appreciated...
- Is anyone able to confirm that the only value for the *unauth_action* argument that has an effect is *pass*, all other values are ignored....

**Comment by bodewig** [comment]
`target_url` will be the URL the user gets forwarded to after the authorization code grant flow has completed. Typically this is not specified explicitly in which case the URL currently visited before starting the flow become the target - https://github.com/zmartzone/lua-resty-openidc/blob/master/lib/resty/openidc.lua#L1363

`unauth_action` is there to decide what to do when `authenticate` is invoked and the user is not known to be authenticated. Currently the only supported values are `'pass'` and "anything else".

* The "anything else" case is the default and means we trigger the authorization code grant flow. 
* `'pass'` means we do not trigger that flow but rather return from `authenticate` with a `nil` token information and an error, the calling code will have to handle the case by itself.

**Comment by bodewig** [comment]
sorry `err` is pretty likely to be `nil` as well if the user is not authenticated. So the calling code detects the situation by looking at the first value returned by `authenticate`.

**Comment by dcs3spp** [comment]
Many thanks @boedwig and @zandbelt for your patience and explanation. Much appreciated 👍

**Comment by zandbelt** [comment]
@dcs3spp it would be awesome if you created a Wiki page with your findings ... :-O

**Comment by dcs3spp** [comment]
Hi @zanbelt,

I have made a small docker-compose demo project available on [gitlab](https://gitlab.com/dcs3spp/angular_keycloak). Feel free to reference on wiki or clone. 

After familiarising myself with lua, I have managed to get the session cookies for identity and access cookies returned when authenticating via keycloak from openresty container @ http://gateway.example.com/api/protected. Hopefully the implementation is correct usage of lua-resty-openidc.....

However, when I try and access the backend API via openresty from the Angular client at http://angular.example.com:4200 there is a CORS cross origin error. 

I supect what is happening is that keycloak is not adding the Access-Control-Allow-Origin header when it is redirecting to login page at Angular client....I suspect that this is maybe what the keycloak javascript adapters do when used between Angular Client and Keycloak directly? 

However, the architecture in the
[gitlab](https://gitlab.com/dcs3spp/angular_keycloak) project is using the gateway as a keycloak client.
